File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_tree/P912_SortAnArray.py
# author: tobebetter9527
# since: 2024/3/11 8:41
from typing import List


class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        if not nums:
            return None
        length = len(nums)
        self.temp = [0 for i in range(length)]
        self.mergeSort(nums, 0, length - 1)
        return nums

    def mergeSort(self, nums, left, right):
        if left == right:
            return
        mid = left + (right - left) // 2
        self.mergeSort(nums, left, mid)
        self.mergeSort(nums, mid + 1, right)
        self.merge(nums, left, mid, right)

    def merge(self, nums, left, mid, right):
        # first copy
        for i in range(left, right + 1):
            self.temp[i] = nums[i]
        # sort the Array
        i, j = left, mid + 1
        idx = left
        while i <= mid and j <= right:
            if self.temp[i] < self.temp[j]:
                nums[idx] = self.temp[i]
                i += 1
            else:
                nums[idx] = self.temp[j]
                j += 1
            idx += 1
        while i <= mid:
            nums[idx] = self.temp[i]
            i += 1
            idx += 1
        while j <= right:
            nums[idx] = self.temp[j]
            j += 1
            idx += 1

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/binary_tree/P912_SortAnArray.java
package com.dsa.leetcode.labuladong.binary_tree;

public class P912_SortAnArray {
    private int[] temp;

    public int[] sortArray(int[] nums) {
        if (nums == null || nums.length == 0) {
            return nums;
        }
        int len = nums.length;
        temp = new int[len];
        // copy
        mergeSort(nums, 0, len - 1);
        return nums;
    }

    private void mergeSort(int[] nums, int left, int right) {
        if (left == right) {
            return;
        }
        int mid = left + (right - left) / 2;
        mergeSort(nums, left, mid);
        mergeSort(nums, mid + 1, right);
        merge(nums, left, mid, right);
    }

    private void merge(int[] nums, int left, int mid, int right) {
        for (int i = left; i <= right; i++) {
            temp[i] = nums[i];
        }

        int i = left;
        int j = mid + 1;
        int idx = left;
        while (i <= mid && j <= right) {
            nums[idx++] = temp[i] < temp[j] ? temp[i++] : temp[j++];
        }
        while (i <= mid) {
            nums[idx++] = temp[i++];
        }
        while (j <= right) {
            nums[idx++] = temp[j++];
        }
    }
}
